source/extensions/config_subscription/grpc/eds_resources_cache_impl.h

```diff
@@ -0,0 +1,48 @@
+#pragma once
+
+#include <memory>
+
+#include "envoy/config/eds_resources_cache.h"
+#include "envoy/event/dispatcher.h"
+#include "envoy/event/timer.h"
+
+#include "absl/container/flat_hash_map.h"
+
+namespace Envoy {
+namespace Config {
+
+class EdsResourcesCacheImpl : public EdsResourcesCache {
+public:
+  EdsResourcesCacheImpl(Event::Dispatcher& main_thread_dispatcher)
+      : dispatcher_(main_thread_dispatcher) {}
+
+  // EdsResourcesCache
+  void setResource(absl::string_view resource_name,
+                   const envoy::config::endpoint::v3::ClusterLoadAssignment& resource) override;
+  void removeResource(absl::string_view resource_name) override;
+  OptRef<const envoy::config::endpoint::v3::ClusterLoadAssignment>
+  getResource(absl::string_view resource_name, EdsResourceRemovalCallback* removal_cb) override;
+  void removeCallback(absl::string_view resource_name,
+                      EdsResourceRemovalCallback* removal_cb) override;
+  uint32_t cacheSizeForTest() const override;
+  void setExpiryTimer(absl::string_view resource_name, std::chrono::milliseconds ms) override;
+  void disableExpiryTimer(absl::string_view resource_name) override;
+
+private:
+  // The value of the map, holds the resource and the removal callbacks.
+  struct ResourceData {
+    envoy::config::endpoint::v3::ClusterLoadAssignment resource_;
+    std::list<EdsResourceRemovalCallback*> removal_cbs_;
```

Comment:
I think std::list probably makes sense here since removal from the list would likely be frequent, but I think if we expect removal to not be a regular occurrence than maybe std::vector is a better choice for better cache locality.

Reply:
Hmmm... good point. I expect the list to be of size 1 or 2 at most. I'm not sure if it will matter that much, but I'm willing to change it to a vector if you think it makes more sense.

Reply:
I think std::vector probably makes more sense, with a call to `reserve(2)` ?  The only benefit of std::list would be frequent random insertion and deletion in a larger list, but with such a small list, that benefit wouldn't exist.

Reply:
I've changed to a vector. I don't think that reserving the size to 2 makes sense here, as the typical use-case is an empty-vector.